meld/cluster_topdown.py

```diff
@@ -0,0 +1,305 @@
+import numpy as np
+import heapq
+from .cluster import sparse_dim_connectivity
+from numba import float64, jitclass, jit
+
+
+@jit(nopython=True)
+def tfce_calc(extent, param_e, upper_limit, h_plus1, lower_limit):
+
+    return np.power(extent, param_e) * \
+                (np.power(upper_limit, h_plus1)
+                 - np.power(lower_limit, h_plus1)) / h_plus1
+
+clus_spec = [
+    ()
+]
+
+
+class Cluster():
+    """ TFCE cluster class
+    """
+    def __init__(self, param_e, param_h):
+        """
+            Initialize empty cluster
+        """
+        self.param_e = param_e
+        self.h_plus1 = param_h + 1
+        # summed TFCE over all cluster members
+        self.accum_val = 0.
+        # extent of cluster; called "totalArea" in Cpp implementation
+        self.extent = 0.
+        # list of cluster members
+        self.members = []
+        # upper limit of TFCE integral; called "lastVal" in Cpp implementation
+        self.upper_limit = 0.
+
+    def update(self, lower_limit):
```

Comment:
Can you specify jit (which I assume means precompile the function) for class member functions?  These get called a lot, and in c/c++, function calls are fairly cheap.

Reply:
The jit decorator specifies that those functions will be compiled and optimized with LLVM the first time they're called. It comes from Numba, https://numba.pydata.org/.

Since the ```self``` object has to be passed and Numba doesn't know how to deal with the python class here, just putting jit here doesn't work. I tried wrapping the entire cluster class as a precompiled class, but that caused getting weird intermittent reference errors after manipulating the cluster objects with the rest of the code. I tried to deal with that by precompiling more of the functions that manipulate the cluster objects but Numba's class implementation really just wraps the class's init function, so other function in the program couldn't take the class as input. Finally, I tried doing away with the class entirely and just keeping lists of everything, but numba doesn't support lists of lists. Now I'm trying Cython instead of Numba, but speeding up this code has been a bit of a rabbit hole.